smart-contract/contracts/YourNftToken.sol

```diff
@@ -37,60 +33,52 @@ contract YourNftToken is ERC721, Ownable, ReentrancyGuard {
     uint256 _maxSupply,
     uint256 _maxMintAmountPerTx,
     string memory _hiddenMetadataUri
-  ) ERC721(_tokenName, _tokenSymbol) {
+  ) ERC721A(_tokenName, _tokenSymbol) {
     cost = _cost;
     maxSupply = _maxSupply;
     maxMintAmountPerTx = _maxMintAmountPerTx;
     setHiddenMetadataUri(_hiddenMetadataUri);
   }
 
   modifier mintCompliance(uint256 _mintAmount) {
-    require(_mintAmount > 0 && _mintAmount <= maxMintAmountPerTx, "Invalid mint amount!");
-    require(supply.current() + _mintAmount <= maxSupply, "Max supply exceeded!");
+    require(_mintAmount > 0 && _mintAmount <= maxMintAmountPerTx, 'Invalid mint amount!');
+    require(totalSupply() + _mintAmount <= maxSupply, 'Max supply exceeded!');
     _;
   }
 
   modifier mintPriceCompliance(uint256 _mintAmount) {
-    require(msg.value >= cost * _mintAmount, "Insufficient funds!");
+    require(msg.value >= cost * _mintAmount, 'Insufficient funds!');
     _;
   }
 
-  function totalSupply() public view returns (uint256) {
-    return supply.current();
-  }
-
   function whitelistMint(uint256 _mintAmount, bytes32[] calldata _merkleProof) public payable mintCompliance(_mintAmount) mintPriceCompliance(_mintAmount) {
     // Verify whitelist requirements
-    require(whitelistMintEnabled, "The whitelist sale is not enabled!");
-    require(!whitelistClaimed[msg.sender], "Address already claimed!");
+    require(whitelistMintEnabled, 'The whitelist sale is not enabled!');
+    require(!whitelistClaimed[msg.sender], 'Address already claimed!');
```

Comment:
```
This contract is only required for intermediate, library-like contracts.
```

That makes me feel that this should be meant for contract libs like `ERC721A` or even the original `ERC721`, while in our case we are not providing users with a library. `YourNftToken.sol` is your final contract.

To my understanding that interface should be used so people will be able override those functions and implement custom behaviors.
Let's say `ERC721A` uses `msg.sender` inside a function which cannot be overridden, then you wouldn't be able to change that unless you take their code and copy-paste it.
In order to use it as a library (e.g. extending it in your contract) they have to use this so you will be able to override `_msgSender()` if needed.

What do you think?

Reply:
yep makes sense! I brought it up too cause I do see some contracts use `_msgSender()` even though its not a  contract lib.

You're the expert here 🙂 so whatever you think is best

Reply:
Nono, that's fine I think it shouldn't make much difference here since people have direct access to the code so they can change it directly if needed, but maybe we can use `_msgSend()` for consistency if it doesn't have any impact on gas efficiency. 😉

Reply:
great points!

Reply:
No impact on gas usage, so I decided to include this change. Thank you 😉

Reply:
that's great! Nice that you were able to determine best direction 🙂